proposals/NNNN-spirv-input-builtin.md

```diff
@@ -0,0 +1,157 @@
+
+## Introduction
+
+HLSL has semantic input/outputs parameters used to carry information
+to and from a shader, e.g: the system semantic `SV_GroupID` or `MY_SEMANTIC`,
+a user-defined semantic.
+
+In SPIR-V, those are translated to `Input` and `Output` variables,
+with either a `BuiltIn` or `Location` decoration.
+
+Example:
+
+```hlsl
+float4 vsmain(float4 Position : POSITION) : SV_POSITION
+{
+  return Position;
+}
+```
+
+Both `POSITION` and `SV_POSITION` will be represented as an `OpVariable`:
+
+- `POSITION` will have the `Input` storage class.
+- `SV_POSITION` will have the `Output` storage class.
+
+- `POSITION` will be decorated with `Location 0`.
+- `SV_POSITION` will be decorated with `BuiltIn Position`.
+
+In addition, users can define their own SPIR-V built-ins using inline SPIR-V:
+
+```hlsl
+[[vk::ext_builtin_input(/* NumWorkGroups */ 24)]]
+static const uint3 myBuiltIn;
+```
+
+This document explain how we plan on implementing those in Clang/LLVM.
+
+## Proposed Solution
+
+## Frontend changes
+
+Global variables marked with Inline SPIR-V will be marked using two new
+attributes:
+- `HLSLInputBuiltin`
+- `HLSLOutputBuiltin`
+
+In addition, a new address space will be added:
+- `vulkan_private`
+
+The TD file will attach each attribute to a `SubjectList` with the following
+constraints:
+
+```
+HLSLInputBuiltin:  S->hasGlobalStorage() &&
+                   S->getStorageClass()==StorageClass::SC_Static &&
+                   S->getType().isConstQualified()
+
+HLSLOutputBuiltin: S->hasGlobalStorage() &&
+                   S->getStorageClass()==StorageClass::SC_Static &&
+                   !S->getType().isConstQualified()
+
+def HLSLVkExtBuiltinInput: InheritableAttr {
+  let Spellings = [CXX11<"vk", "ext_builtin_input">];
+  let Args = [IntArgument<"BuiltIn">];
+  let Subjects = SubjectList<[HLSLInputBuiltin], ErrorDiag>;
+  let LangOpts = [HLSL];
+}
+
+def HLSLVkExtBuiltinOutput: InheritableAttr {
+  let Spellings = [CXX11<"vk", "ext_builtin_output">];
+  let Args = [IntArgument<"BuiltIn">];
+  let Subjects = SubjectList<[HLSLOutputBuiltin], ErrorDiag>;
+  let LangOpts = [HLSL];
+}
+```
+
+When this attribute is encountered, several changes will occur:
+- The address space will be set to `vulkan_private`
+- a constructor will be added.
+- a destructor will be added.
+
+The construction will call a target-specific intrinsic: `spv_load_builtin`.
+This builtin takes the SPIR-V built-in ID as parameter, and returns the
+value of the corresponding built-in.
+
+The destructor will call another intrinsic: `spv_store_builtin`. This builtin
+takes the SPIR-V built-in ID, as well as a value to store as parameter.
+
+And lastly, to support input parameters with semantics, the same intrinsic
+will be called in the generated entrypoint wrapper function.
+
+Because the 2 new intrinsics will be marked as `MemWrite` or `MemRead`, LLVM
+won't be able to optimize them away, and the initial/final load/store will
+be kept.
```

Comment:
When looking at what DXC does, I see it loads/stores only under control flow, when that's the way it's used in the entry.  I think this proposal is written to imply that it will not match that behavior.  I imagine that an unused output will have an unconditional write at the end of every entry that has such a global declaration in the file.

This proposal offers no mechanism for guaranteeing accesses under control flow remain under that control flow, and no additional accesses are added to the built-ins as a side effect of some implementation detail, for either load or store.

Reply:
Splitting 2 issues here:
 - unused input/output not being discarded/removed
 - control flow access
 
@s-perron to double check, but AFAIK in SPIR-V, The Built-in value is ONLY visible to the executing lane. Meaning having a load/store moved outside of a given control-flow wouldn't impact the behavior of the shader.
(Specifically talking about the control flow bit, non-discarded unused built ins is another issue)